bandersnatch_vrfs/src/lib.rs

```diff
@@ -83,132 +77,148 @@ type ThinVrf = dleq_vrf::ThinVrf<E>;
-        ring_prover: &RingProver
-    ) -> RingVrfSignature<N>
-    {
-        assert_eq!(ios.len(), N);
-        let blinding_base = ring_prover.piop_params().h;
-        let (signature,secret_blinding) = pedersen_vrf(blinding_base).sign_pedersen_vrf(t, ios, None, &self.0);
-        let preoutputs = vrf::collect_preoutputs_array(ios);
-        let ring_proof = ring_prover.prove(secret_blinding.0[0]);
-        RingVrfSignature { preoutputs, signature, ring_proof, }
+impl ArkScaleMaxEncodedLen for RingVrfProof {
+    fn max_encoded_len() -> usize {
+        <PedersenVrfProof as ArkScaleMaxEncodedLen>::max_encoded_len()
+        + 4096  // TODO: How large is RingProof, Sergey?
     }
 }
 
+// #[derive(Debug,Clone)]
+pub struct RingVerifier(pub ring::RingVerifier);
 
-pub const PUBLIC_KEY_LENGTH: usize = 33;
-pub type PublicKeyBytes = [u8; PUBLIC_KEY_LENGTH];
+pub type RingVrfSignature<const N: usize> = dleq_vrf::VrfSignature<RingVerifier,N>;
 
-#[derive(Debug,Clone,CanonicalSerialize,CanonicalDeserialize)]
-pub struct PublicKey(pub dleq_vrf::PublicKey<E>);
+impl EcVrfVerifier for RingVerifier {
+    type H = E;
+    type VrfProof = RingVrfProof;
+    type Error = SignatureError;
 
-impl PublicKey {
-    pub fn serialize(&self) -> PublicKeyBytes {
-        let mut bytes = [0u8; PUBLIC_KEY_LENGTH];
-        self.serialize_compressed(bytes.as_mut_slice())
-        .expect("Curve needs more than 33 bytes compressed!");
-        bytes
-    }
+    fn vrf_verify_detached<'a>(
+        &self,
+        t: impl IntoTranscript,
+        ios: &'a [VrfInOut],
+        signature: &RingVrfProof,
+    ) -> Result<&'a [VrfInOut],Self::Error> {
+        let ring_verifier = &self.0;
+        let blinding_base = ring_verifier.piop_params().h;
+        pedersen_vrf(blinding_base).verify_pedersen_vrf(t,ios.as_ref(),&signature.dleq_proof) ?;
 
-    pub fn deserialize(reader: &[u8]) -> Result<Self, SerializationError> {
-        Self::deserialize_compressed(reader)
+        let key_commitment = signature.dleq_proof.as_key_commitment();
+        match ring_verifier.verify_ring_proof(signature.ring_proof.clone(), key_commitment.0.clone()) {
+            true => Ok(ios),
+            false => Err(SignatureError::Invalid),
+        }
     }
 }
 
-ark_scale::impl_scale_via_ark!(PublicKey);
-
-impl MaxEncodedLen for PublicKey {
-    #[inline]
-    fn max_encoded_len() -> usize {
-        crate::PUBLIC_KEY_LENGTH
+impl RingVerifier {
+    pub fn verify_ring_vrf<const N: usize>(
+        &self,
+        t: impl IntoTranscript,
+        inputs: impl IntoIterator<Item = impl IntoVrfInput<E>>,
+        signature: &RingVrfSignature<N>,
+    ) -> Result<[VrfInOut; N],SignatureError>
+    {
+        self.vrf_verify(t, inputs, signature)
     }
 }
 
 
-// pub type ThinVrfSignature<const N: usize> = dleq_vrf::VrfSignature<ThinVrf>;
- 
-// VrfSignature
+// #[derive(Clone)]
+pub struct RingProver<'a> {
+    pub ring_prover: &'a ring::RingProver,
+    pub secret: &'a SecretKey,
+}
```

Comment:
We could add parameters here now, like maybe the `blinding_base` @swasilyev just removed from `ring::Ring*`, but that one makes more sense as a global constant imho.